src/charm.py

```diff
@@ -291,6 +303,98 @@ def _on_db_router_relation_changed(self, event: RelationChangedEvent) -> None:
+        app_relation_databag = event.relation.data[self.app]
+
+        if unit_relation_databag.get("password"):
+            # Test if relation data is already set
+            # and avoid overwriting it
+            logger.warning("Data for shared-db already set.")
+            self.unit.status = ActiveStatus()
+            return
+
+        # retrieve data from the relation databag
+        # Abstracted is the fact that it is received from the leader unit of the related app
+        requires_relation_databag = event.relation.data[event.unit]
+        database_name = requires_relation_databag.get("database")
+        database_user = requires_relation_databag.get("username")
+
+        if not database_name or not database_user:
+            # Cannot create scoped database without credentials
+            logger.warning(
+                "Missing information for shared-db relation to create a database and scoped user"
+            )
+            self.unit.status = ActiveStatus()
+            return
+
+        password = generate_random_password(PASSWORD_LENGTH)
+
+        try:
+            self._mysql.create_application_database_and_scoped_user(
+                database_name, database_user, password
+            )
+
+            # set the relation data for consumption
+            cluster_primary = str(self.model.get_binding(PEER).network.bind_address)
+
+            unit_relation_databag["db_host"] = cluster_primary
+            # Database port is static in legacy charm
+            unit_relation_databag["db_port"] = "3306"
+            # Wait timeout is a config option in legacy charm
+            # defaulted to 3600 seconds
+            unit_relation_databag["wait_timeout"] = "3600"
+            unit_relation_databag["password"] = password
+
+            unit_names = " ".join([unit.name for unit in event.relation.units])
+            unit_relation_databag["allowed_units"] = unit_names
+
+            # store username for relation in app databag
+            # this is used to remove the user when the relation is broken
+            app_relation_databag[f"relation_id_{event.relation.id}_db_user"] = database_user
+
+        except (
+            MySQLClientError,
+            MySQLCreateApplicationDatabaseAndScopedUserError,
+        ):
+            self.unit.status = BlockedStatus("Failed to initialize shared_db relation")
+            return
+
+        self.unit.status = ActiveStatus()
+
+    def _on_shared_db_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle the departure of legacy shared_db relation.
+
+        Remove user created for the relation but keep the database.
+        """
+        if not self.unit.is_leader():
+            return
+
+        app_relation_databag = event.relation.data[self.app]
+        username = app_relation_databag.get(f"relation_id_{event.relation.id}_db_user")
+
+        if not username:
+            # Can't do much if we don't have the username
+            return
```

Comment:
Should this be logged in some way? If there's no username where we expect one to be, it might be a bad idea to fail quietly.